stories/dashboard-details.stories.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Dashboard Details page stories.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { storiesOf } from '@storybook/react';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_NAME as CORE_SITE } from '../assets/js/googlesitekit/datastore/site/constants';
+import { STORE_NAME as CORE_USER } from '../assets/js/googlesitekit/datastore/user/constants';
+import { WithTestRegistry } from '../tests/js/utils';
+import DashboardDetailsApp from '../assets/js/components/dashboard-details/dashboard-details-app';
+import DashboardDetailsEntityNotFoundView from '../assets/js/components/dashboard-details/DashboardDetailsEntityNotFoundView';
+
+storiesOf( 'Dashboard Details', module )
+	.add( 'Existing Entity', () => {
+		global.featureFlags = {
+			widgets: {
+				pageDashboard: {
+					enabled: false,
+				},
+			},
+		};
```

Comment:
We should be mindful to prevent stories from having side-effects that persist between stories.

Ideally, globals would be reset between each story like we do in other tests but I'm not sure we can do that globally with Storybook. We have some stories that do this already using decorators, or manually in the Story function but I think we can handle this a bit better with a wrapper component.

Let's introduce a wrapper component for forcing specific global values before rendering and restoring the previous values on unmount.

Essentially something like this:

```js
function WithGlobals( { children, globals } ) {
	const previousValues = _.pick( global, globals );
	const restoreGlobals = useCallback( () => Object.assign( global, previousValues ), [] );

	useEffect( () => {
		Object.assign( global, globals )
		return () => restoreGlobals(),
	}, []);

	return children;
}
```

Alternatively, maybe something like this might be a bit better:

```js
function WithEffects( { children, setUp = () => {}, tearDown = () => {} } ) {
	useEffect( () => {
		setUp()
		return tearDown,
	}, []);

	return children;
}
// and then WithGlobals could be based on this
function WithGlobals( { children, globals } ) {
	const previousValues = _.pick( global, globals );
	const restoreGlobals = useCallback( () => Object.assign( global, previousValues ), [] );

	return <WithEffects
		setUp={ () => Object.assign( global, globals ) }
		tearDown={ restoreGlobals }
		children={ children }
	/>
}
```

That way components needing more complex setup/teardown could use the lower-level `WithEffects`.

What do you think?

Reply:
I like this idea, but for some reason, I can't make it work. `useEffect` hook isn't being triggered and I can't figure out why :/

Reply:
I'll take a closer look. I didn't test my suggestion there but it seems like it should work.

Reply:
I'm seeing a similar issue, and I'm not yet sure why. It might be worth exploring that in another issue to not block this—maybe storing the feature flags in context or something less fragile could be a good solution going forward…

Reply:
Ah yes, of course 😄  The issue is `<DashboardDetailsApp />` will render before the `<WithGlobals>` function. It's not exactly that the `useEffect` _doesn't run_, but that the innermost components will get rendered first, and thus error before the `global.featureFlags` variables are set. If you replace the approach with something that defines the globals before rendering, but _removes_ them on unmount, you can see it works:

```js
function ResetFeatureFlagsOnUnmount( { children } ) {
	useEffect( () => {
		return () => {
			delete global.featureFlags;
		};
	}, [] );

	return children;
}

storiesOf( 'Dashboard Details', module )
	.add( 'Existing Entity', () => {
		const setupRegistry = ( registry ) => {
			provideUserAuthentication( registry );
			provideSiteInfo( registry, {
				currentEntityTitle: 'Test Page',
				currentEntityURL: 'https://example.com/test-page',
			} );
		};
		global.featureFlags = {
			widgets: {
				pageDashboard: {
					enabled: false,
				},
			},
		};

		return (
			<ResetFeatureFlagsOnUnmount>
				<WithTestRegistry callback={ setupRegistry }>
					<DashboardDetailsApp />
				</WithTestRegistry>
			</ResetFeatureFlagsOnUnmount>
		);
	} )
	.add( 'Not Found Entity', () => (
		<DashboardDetailsEntityNotFoundView permalink="https://example.com/test-page" />
	) );
```

will work as expected:

![2020-10-08 15 32 21](https://user-images.githubusercontent.com/90871/95472997-8dd36180-097b-11eb-9242-31a266c6870c.gif)

-----

https://imkev.dev/react-rendering-order is a handy blog post to understand component rendering. It's a bit counter-intuitive, but explains why this is happening 😄

I think having a better way of dealing with globals is good, though maybe it means we should be injecting featureFlags into context to make them easier to mock. We should do that in a separate issue though, so I think this PR is good-to-go.

Reply:
Thanks @tofumatt – I think this might be good to add to the global decorator so we don't have to do it for every story, what do you think?
https://github.com/google/site-kit-wp/blob/71c25b424182cfef6ea115ae1c77b17bdf488a6a/.storybook/config.js#L105-L110

Reply:
I think adding a Storybook entry for this goes a bit too far here - the UI here is not new, so it's strictly not responsibility of this PR to add it. Because of the involved complexity, both with the above discussion as well as we'd need to provide data for all the widgets to avoid the load flashing, I think it should be removed from this PR.